src/amo/withExperiment.js

```diff
@@ -141,63 +154,80 @@ export const withExperiment = ({
   invariant(defaultVariants, 'variants is required');
 
   class WithExperiment extends React.Component<withExperimentInternalProps> {
-    experimentCookie: string | void;
-
     static defaultProps = {
       _config: config,
       _getVariant: getVariant,
+      _isExperimentEnabled: isExperimentEnabled,
       id: defaultId,
       variants: defaultVariants,
     };
 
     static displayName = `WithExperiment(${getDisplayName(WrappedComponent)})`;
 
-    constructor(props: withExperimentInternalProps) {
-      super(props);
+    getExperiments() {
+      const { cookies } = this.props;
+
+      return cookies.get(EXPERIMENT_COOKIE_NAME) || {};
+    }
+
+    cookieIncludesExperiment(registeredExperiments: RegisteredExpermients) {
+      const { id } = this.props;
+
+      return Object.keys(registeredExperiments).includes(id);
+    }
 
-      if (!this.isExperimentEnabled()) {
-        log.debug(`Experiment "${props.id}" is not enabled by config.`);
-        return;
+    render() {
+      const {
+        _config,
+        _getVariant,
+        _isExperimentEnabled,
+        cookies,
+        id,
+        variants,
+        ...props
+      } = this.props;
+
+      const isEnabled = _isExperimentEnabled({ _config, id });
```

Comment:
We shouldn't have this logic in a `render()` method. Could you move it to the `constructor` instead? The main reason is that `render()` can be called multiple times.

Reply:
Moved it back into `constructor`.